File: current_session/python/46.py
"""
see https://leetcode.com/problems/permutations/
"""
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if not nums:
            return [[]]
        if len(nums) == 1:
            return [[n] for n in nums]
        res = []
        for i in range(len(nums)):
            tmp = self.permute([a for a in nums if a!=nums[i]])
            for t in tmp:
                res.append([nums[i]]+t)
        return res

print(Solution().permute([1,2,3,4]))
print(Solution().permute([1,2,3]))
print(Solution().permute([1]))
print(Solution().permute([]))

File: current_session/c++/46.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> permute(vector<int>& nums) {
    vector<vector<int> > res;
    if (nums.size() == 1) {
        res.push_back(nums);
        return res;
    }
    for (int i=0; i<nums.size(); ++i) {
      swap(nums[i], nums[nums.size()-1]);
      int x = nums[nums.size()-1];
      nums.pop_back();
      vector<vector<int> > tmp = permute(nums);
      nums.push_back(x);
      swap(nums[i], nums[nums.size()-1]);
      for (auto sol:tmp) {
        sol.push_back(x);
        res.push_back(sol);
      }
      // res.push_back(tmp);
    }
    return res;
  }
};